### ISSUE TITLE
Unable to pass workdir 

### ISSUE THREAD

**Issue by ssala7** [jenkinsci/docker-plugin#561]
Unable to pass workdir,  i want to pass the workdir / -w option when we create a docker using docker-plugin

Available Options:
=============
Image name |   |  
-- | -- | --
  |   |  
  | Command |   |  
  |   |  
  | Hostname |   |  
  |   |  
  | Container name |   |  
  |   |  
  |   |   |  
  | Environment variables |   |  
  |   |  
  | /n\|/r separated list of environment to be started. Example: var1=value1 /n\|/r var2=value2 |  
  |   | Check syntax |  
  |   |  
  | Links |   |  
  |   |  
  | Comma separated list of links. Example: name1:alias1,name1:alias2 |  
  | Exposed ports |   |  
  |   |  
  | Comma separated list of ports to be exposed. Example: 9000/tcp |  
  | CPU shares |   |  
  |   |  
  | CPU shares (relative weight) |  
  | Memory limit |   |  
  |   |  
  | Memory limit (format: number{optional unit}, where unit = b, k, m or g). Example: 1024m |  
  | DNS |   |  
  |   |  
  | Option --dns:Comma separated list of links. Example: dns1,dns2,... |  
  | Extra Hosts |   |  
  |   |  
  | Option --add-host:Comma separated list of add-Host.Add hostnames to /etc/hosts in the container. Example: central.maven.org:185.31.18.209,repo1.maven.org:185.31.18.209,... |  
  | Publish all ports |   |  
  |   |  
  |   |  
  | Port bindings |   |  
  |   |  
  |   |  
  |   | Check syntax |  
  |   |  
  | Bind mounts |   |  
  |   |  
  |   |  
  |   | Check syntax |  
  |   |  
  | Extended privileges |   |  
  |   |  
  |   |  
  | always restart |

**Comment by ndeloof** [comment]
Can you please describe why you want to change workdir as defined by image ?

**Comment by guessi** [comment]
+1 same problem, seeking for fixes,

with `docker.image(xxxxxx).inside('-w my-workdir') { ... }`
docker plugins will always append `-w ${WORKSPACE}` at the `END`, that's the issue,

latter `-w` wins, override the `-w my-workdir` in `inside()` block,

related issue, https://github.com/jenkinsci/docker-workflow-plugin/pull/123

**Comment by ndeloof** [comment]
jenkinsci/docker-workflow-plugin#123 is for another use-case, which is support for `dir` within pipeline. Using a docker-plugin's node this isn't an issue as you can change current dir for jenkins tasks at any time.
So my question : why do you need to configure your agent to run from image X with another working dir but the one it has been designed to ? All process ran by Jenkins should run with current workspace as working dir, don't they ?

**Comment by guessi** [comment]
yes, I agree all source changes should be happened under ${WORKSPACE}, but not for `WORKDIR`

take GOLANG as example, it need to switch to ${GOPATH}/, and it can build/run their code, so it is still required for this use case.

**Comment by ndeloof** [comment]
golang image is a good sample. Please remember docker-plugin require you to provide a Docker image designed as a jenkins build agent. So builds will run under workspace, as for any jenkins agent. The fact golang image uses a pre-configured GOPATH=/go just means you have to setup your job to override GOPATH.

**Comment by guessi** [comment]
yeah, but it would be better to have customized `WORKDIR`

logic might be,
```
if ("-w option found in inside() block" && "-w existed exactly once") {
  set customized WORKDIR as defined in inside() block
} else {
  set WORKDIR as ${WORKSPACE}
}
```

**Comment by ndeloof** [comment]
this would break many other tool integrations
the fact a docker container can be ran this way does not mean we have to change the implicit contract for a jenkins agent with the other APIs. Just write your job for arbitrary jenkins agent, and benefit docker for elasticity and isolation.

**Comment by guessi** [comment]
@ndeloof got it, thanks for reply

**Comment by nabheet** [comment]
```So my question : why do you need to configure your agent to run from image X with another working dir but the one it has been designed to ? All process ran by Jenkins should run with current workspace as working dir, don't they ?```

Because I need to run multiple images to run tests for the one that is currently being built.

I think this is a bug and should be reopened. At the very least you should provide the ability to override the workdir.

**Comment by robindegen** [comment]
I'm also running into this issue. I can answer this question quite simply: Docker on Windows.
My windows docker now fails with:

```
Caused: java.io.IOException: Cannot run program "cmd" (in directory "/var/jenkins_home/workspace/zlib_windows_vc2019"): error=2, No such file or directory
```

Which makes sense because obviously that isn't a valid windows path. I'm using traditional "freestyle" jenkins builds. I have no idea how to fix this without configuring a specific home directory per docker cloud connection

**Comment by bluejack2000** [comment]
I have a big problem too while running jenkins on Windows and using linux docker images. It attempts to map windows paths into docker containers which causes the docker command to fail. The problem is that it uses the SAME path inside the docker container. Windows paths do not work inside linux.

**Comment by MarioArgus** [comment]
I have same issue as @bluejack2000, error returned is "...working directory is invalid.. it needs to be an absolute path"

**Comment by pjdarton** [comment]
Folks - this issue is closed.

If you're using a pipeline and want to use docker, that's documented in https://jenkins.io/doc/book/pipeline/docker/ and you'll be using the https://plugins.jenkins.io/docker-workflow plugin, which was designed for pipeline usage, and not this one.
*This* plugin was not designed to be controlled from pipelines.
See also: last paragraph on https://github.com/jenkinsci/docker-plugin/blob/master/README.md

If you're not controlling your docker stuff from pipelines but are, instead, using this plugin simply to provide docker containers from templates to run your builds (which might be pipelines or old-style builds) then by all means raise a bug and include screenshots and stacktraces etc, explaining exactly what happens, how to reproduce the issue, and what you'd like it to do instead ... but read https://github.com/jenkinsci/docker-plugin/blob/master/README.md first.
For example, with _this_ plugin, when you tell docker to mount `Volumes` from the docker host into containers, you have a choice of specifying both the dir on the docker host's filesystem and the dir within the container where you'd like it to appear, so that's all under user-control.
This plugin is dependent on the docker-java library and I guess it's possible that it might be that code that's complaining (which would mean that such a bug would have to be fixed in that library first, and then this plugin updated to use the newer library), but (as with most issues) we'd need a stacktrace to be able to tell.

TL;DR: This issue is closed; please start over, making sure you're starting in the right place with the right information.
